File: merge_sort.py
def combine(array, low, mid, high):
    i, j, temp = low, mid + 1, list()
    while i <= mid and j <= high:
        if array[i] <= array[j]:
            temp.append(array[i])
            i += 1
        else:
            temp.append(array[j])
            j += 1
    while i <= mid:
        temp.append(array[i])
        i += 1
    while j <= high:
        temp.append(array[j])
        j += 1
    array[low:high + 1] = temp


def split(array, low, high):
    if low < high:
        mid = (low + high) / 2
        split(array, low, mid)
        split(array, mid + 1, high)
        combine(array, low, mid, high)

array = [2, 5, 1, 6, 0, 1, 6, 21, 4, 32]
split(array, 0, len(array) - 1)
for i in array:
    print i

File: merge_sort.cpp
#include <iostream>
#define size 8

void combine(int *array, int low, int mid, int high){
    int i = low, j = mid + 1, k = low;
    int temp[size];
    j = mid+1;
    while(i <= mid && j <= high)
        if(array[i]<array[j])
            temp[k++] = array[i++];
        else
            temp[k++] = array[j++];
    while(i<=mid)
        temp[k++] = array[i++];
    while(j<=high)
        temp[k++] = array[j++];
    for(k=low;k<=high;++k)
        array[k] = temp[k];
}
void split(int *array, int low, int high){
    if(low<high){
        int mid = (low+high)/2;
        split(array, low, mid);
        split(array, mid+1, high);
        combine(array, low, mid, high);
    }
}
using namespace std;
int main()
{
    int array[] = {3,2,7,1,3,4,7,0};
    split(array, 0, size-1);
    for (auto i : array ) {
        cout << i << endl;
    }
    return 0;
}
